examples/custom-client-cxx/cli.cpp

```diff
@@ -0,0 +1,150 @@
+#include <thread>
+
+#include <boost/process.hpp>
+#include <boost/log/core.hpp>
+#include <boost/log/trivial.hpp>
+#include <boost/log/expressions.hpp>
+
+#include "aktualizr-lite/api.h"
+#include "cli.h"
+#include "cli/cli.h"
+
+#define LOG_INFO    BOOST_LOG_TRIVIAL(info)
+#define LOG_WARNING BOOST_LOG_TRIVIAL(warning)
+#define LOG_ERROR   BOOST_LOG_TRIVIAL(error)
+
+int cmd_check(std::string local_repo_path) {
+  boost::log::core::get()->set_filter(boost::log::trivial::severity >= boost::log::trivial::info);
+  bool is_local_update{false};
+  LocalUpdateSource local_update_source;
+
+  std::vector<boost::filesystem::path> cfg_dirs;
+  auto env{boost::this_process::environment()};
+  if (env.end() != env.find("AKLITE_CONFIG_DIR")) {
+    cfg_dirs.emplace_back(env.get("AKLITE_CONFIG_DIR"));
+  } else {
+    cfg_dirs = AkliteClient::CONFIG_DIRS;
+  }
+
+  if (!local_repo_path.empty()) {
+    local_update_source = {
+      local_repo_path + "/tuf",
+      local_repo_path + "/ostree_repo",
+      local_repo_path + "/apps"
+    };
+    is_local_update = true;
+    LOG_INFO << "Offline mode. Updates path=" << local_repo_path;
+  } else {
+    LOG_INFO << "Online mode";
+  }
+
+  std::unique_ptr<AkliteClient> client;
+  try {
+    client = std::make_unique<AkliteClient>(cfg_dirs);
+  } catch (const std::exception& exc) {
+    LOG_ERROR << "Failed to initialize the client: " << exc.what();
+    return EXIT_FAILURE;
+  }
+
+  auto ret = aklite::cli::CheckIn(*client, is_local_update ? &local_update_source : nullptr);
+  return aklite::cli::IsSuccessCode(ret)? EXIT_SUCCESS : EXIT_FAILURE;
+}
+
+int cmd_pull(std::string target_name, std::string local_repo_path) {
+  boost::log::core::get()->set_filter(boost::log::trivial::severity >= boost::log::trivial::info);
+  bool is_local_update{false};
+  LocalUpdateSource local_update_source;
+
+  std::vector<boost::filesystem::path> cfg_dirs;
+  auto env{boost::this_process::environment()};
+  if (env.end() != env.find("AKLITE_CONFIG_DIR")) {
+    cfg_dirs.emplace_back(env.get("AKLITE_CONFIG_DIR"));
+  } else {
+    cfg_dirs = AkliteClient::CONFIG_DIRS;
+  }
+
+  if (!local_repo_path.empty()) {
+    local_update_source = {
+      local_repo_path + "/tuf",
+      local_repo_path + "/ostree_repo",
+      local_repo_path + "/apps"
+    };
+    is_local_update = true;
+    LOG_INFO << "Offline mode. Updates path=" << local_repo_path;
+  } else {
+    LOG_INFO << "Online mode";
+  }
+
+  std::unique_ptr<AkliteClient> client;
+  try {
+    client = std::make_unique<AkliteClient>(cfg_dirs);
+  } catch (const std::exception& exc) {
+    LOG_ERROR << "Failed to initialize the client: " << exc.what();
+    return EXIT_FAILURE;
+  }
+
+  auto ret = aklite::cli::Pull(*client, -1, "", InstallMode::All, true, is_local_update ? &local_update_source : nullptr);
+  return aklite::cli::IsSuccessCode(ret)? EXIT_SUCCESS : EXIT_FAILURE;
```

Comment:
Why not return the exit code returned from `aklite::cli::Pull`?

Reply:
Returning the exit code from `aklite::cli` functions directly, while making it clear if it is a success or a failure, and what to do next, sounds a reasonable approach.

I'm a heavy user of the `&&` operator when running multiple commands in shell, so I tend to like having 0 for success whenever possible. But the consistency with the remaining tools, and the fact that the flow should in fact depend on the returned value, are more important.

Do you think it is worth to keep the `IsSuccessCode` function?

Reply:
This is just an example, so I think it is fine to return 0 or 1.
What is an important is to add `switch case` (or if) statement that prints a message that corresponds to the `check/pull/install/run` execution result and in some cases instructs what to do the next.

Whether to return 0 or 1 is up to a customer to decide. I think it depends on a specific use-case.  I can imagine use-case when a custom client triggers another binary/script as a reaction on certain `ret` values (e.g. reboot), so the custom client doesn't need to return specific exit codes, just 0 and 1.

Reply:
I've added switch case logic for every possible return code. Since there is quite a lot of duplication between possible return codes for each command, I've concentrated all of them in a single `printStatus` function. Let me know what you think about it.

Reply:
Looks fine to me.